File: osm_fetch/yaml_parser.py
from enum import Enum
import argparse

from pydantic import BaseModel
from pydantic_yaml import parse_yaml_file_as
from rich.pretty import pprint

class Operation(str, Enum):
    difference = "difference"
    union = "union"

class ConfigLayer(BaseModel):
    name : str
    operation : Operation

class UnsealingConfig(BaseModel):
    layers : list[ConfigLayer]
    city: str

def parse_config(filename):
    return parse_yaml_file_as(UnsealingConfig, filename)

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("file")
    args = parser.parse_args()
    filename = args.file
    config = parse_yaml_file_as(UnsealingConfig, args.file)
    pprint(config)

File: polygon_analyzer/lib/yaml_parser.dart
import 'dart:convert';
import 'dart:io';
import 'package:yaml/yaml.dart';
import 'package:json_annotation/json_annotation.dart';
part 'yaml_parser.g.dart';

enum Operation { difference, union }

@JsonSerializable()
class ConfigLayer {
  String name;
  Operation operation;

  ConfigLayer({required this.name, required this.operation});

  factory ConfigLayer.fromJson(Map<String, dynamic> json) =>
      _$ConfigLayerFromJson(json);

  Map<String, dynamic> toJson() => _$ConfigLayerToJson(this);
}

@JsonSerializable()
class UnsealingConfig {
  List<ConfigLayer> layers;
  String city;

  UnsealingConfig({required this.layers, required this.city});

  factory UnsealingConfig.fromJson(Map<String, dynamic> json) =>
      _$UnsealingConfigFromJson(json);

  Map<String, dynamic> toJson() => _$UnsealingConfigToJson(this);
}

UnsealingConfig parseConfig(String filename) {
  final fileContent = File(filename).readAsStringSync();
  final yamlMap = jsonDecode(jsonEncode(loadYaml(fileContent) as YamlMap))
      as Map<String, dynamic>;

  return UnsealingConfig.fromJson(yamlMap);
}
